### ISSUE TITLE
Universal Links can be dropped by Branch

### ISSUE THREAD

**Issue by welshm** [BranchMetrics/ios-branch-deep-linking#284]
The universal link passed into the application sometimes does not get passed to the Branch handler (initSessionWithLaunchOptions:andRegisterDeeplinkHandler:).

The following race condition exists:

  1. Initializes when the application resumes. This can happen before the universal/deeplink param is set. A init request is sent.

  2. The universal/deeplink is set in BNCPreferenceHelper

  3. The request from (1.) returns and wipes out the universal link in BNCPreferenceHelper - and thus it never gets  reported to the application by the branch handler block.

I tried to reset the session, but because the request from (1.) is still active, it just moves it to the top of the queue.

**Comment by derrickstaten** [comment]
@welshm could you provide (a) the link you're clicking, and (b) the logs from setDebug (https://dev.branch.io/recipes/testing_your_integration/ios/#use-setdebug-to-simulate-fresh-installs)? if you're worried about sharing the latter publicly, please send to support@branch.io. 

Please also include exact steps to reproduce.

That will help us try to reproduce.

Currently I've never seen this and I've done quite a bit of manual testing. I'm looking forward to the logs + steps to reproduce so we can isolate this!

**Comment by welshm** [comment]
I contacted support as we have concerns sharing publicly

**Comment by derrickstaten** [comment]
@welshm I can reproduce this with AirDrop when the app is in the foreground. Unfortunately we currently don't support continueUserActivity while the application is already active and in the foreground. I'm worried that there may not be a good fix that doesn't result in breaking more heavily used uses of continueUserActivity, like Universal Links and Spotlight Search.

Outside of AirDrop, I've been unable to reproduce.

Do you have steps to reproduce that don't involve AirDrop with the app in the foreground?

**Comment by derrickstaten** [comment]
@welshm there is no easy fix here, unfortunately. Long-term, we should probably move away from looking at whether an app is active to looking at whether an app is in the foreground, but that will require a substantial rewiring of our SDK and a lot of testing to make sure we handle all possible scenarios, as it would re-open race conditions that we've addresses previously.

TL;DR the inability to share via AirDrop is now a known issue without easy resolution.

**Comment by welshm** [comment]
This was closed - was it fixed by a change?

**Comment by ahmednawar** [comment]
No, but this will require a large rewrite of the SDK which won't happen anytime soon.

**Comment by aaustin** [comment]
Closing for lack of prevalence.
